### PR TITLE
Remove ja_resource and canary reliance from UserController

### PR DESCRIPTION
**Pull request by snewcomer**
# What's in this PR?
[WIP]
Remove ja_resource from User Controller

## References
Fixes #882

Progress on: #864

### CODE REVIEW

File: lib/code_corps/policy/policy.ex [hunk 1]
```diff
@@ -35,8 +38,9 @@ defmodule CodeCorps.Policy do
     # will never do anything
     #
     # The only solution is to have a catch_all match for the resource being nil, which returns true
+    
+    # NOTE: other tests are using the User policy for the time being.
     def can?(%User{}, _action, nil), do: true
-
     def can?(%User{} = current_user, :update, %User{} = user), do: Policy.User.update?(user, current_user)
```
Comment (snewcomer) [inline lib/code_corps/policy/policy.ex]: NOTE: had to leave these in here b/c other tests not in the `user_controller_test` were depending on this

File: lib/code_corps_web/controllers/user_controller.ex [hunk 2]
```diff
@@ -68,9 +71,9 @@ defmodule CodeCorpsWeb.UserController do
     conn |> json(hash)
   end
 
-  defp login(conn, _opts) do
-    Plug.Conn.register_before_send(conn, &do_login(&1))
-  end
+  # defp login(conn, _opts) do
```
Comment (snewcomer) [inline lib/code_corps_web/controllers/user_controller.ex]: @begedin any idea why commenting these lines (and the plug :login) above out allows for a passing test suite?  Otherwise I get this [line](https://github.com/code-corps/code-corps-api/blob/develop/test/lib/code_corps/github/installation_test.exs#L46) and this [line](https://github.com/code-corps/code-corps-api/blob/develop/test/lib/code_corps/github/installation_test.exs#L75) failing

Comment (begedin) [reply lib/code_corps_web/controllers/user_controller.ex]: I'm not sure, but I would expect the plug to cause things to fail either way. It was added mainly to get around some ja_resource quirks.

What I would do is remove `plug :login when action in [:create]`, as well as `defp login` and `defp do_login` completely.

Then, in `create/2`, I would add a `|> Guardian.Plug.sign_in(user)` between `conn` and `|> put_status(:created)`, or really, anywhere in that chain.

That would be the most explicit and easiest to read approach for dealing with it anyway.

File: lib/code_corps/policy/policy.ex (hunk 1 above)
Comment (begedin) [reply lib/code_corps/policy/policy.ex]: I understand why line 44 needs to stay, but are you sure about 45? Which tests fail if you remove it? Assuming you add the "fixes" i described below.

**Review by begedin — CHANGES_REQUESTED** [review]
This generally looks good. Ping me for another review when you try the advice I gave.

File: lib/code_corps_web/controllers/user_controller.ex (hunk 2 above)
Comment (snewcomer) [reply lib/code_corps_web/controllers/user_controller.ex]: @begedin is this b/c the user is signed in automatically after creating a user?

File: lib/code_corps_web/controllers/user_controller.ex [hunk 3]
```diff
@@ -1,40 +1,42 @@
...
+    end
   end
 
-  def handle_create(_conn, attributes) do
-    %User{} |> User.registration_changeset(attributes)
+  @spec create(Conn.t, map) :: Conn.t
+  def create(%Conn{} = conn, %{} = params) do
+    with {:ok, %User{} = user} <- %User{} |> User.registration_changeset(params) |> Repo.insert
+    do
+      conn |> Plug.Conn.fetch_session |> Guardian.Plug.sign_in(user) |> put_status(:created) |> render("show.json-api", data: user)
```
Comment (snewcomer) [inline lib/code_corps_web/controllers/user_controller.ex]: @begedin so this worked.

Comment (snewcomer) [reply lib/code_corps_web/controllers/user_controller.ex]: Although a bit lost on why this is needed as compared to before these changes.

Comment (begedin) [reply lib/code_corps_web/controllers/user_controller.ex]: The `login` function was a plug. It automatically ran before performing the `create` action and told the conn that, after performing `create`, it needs to run the `do_login` function.

The `do_login` function then signed the `user` into the `conn` if the `create` action was successful.

Basically, from a user's perspective, "i want to be signed in when I register".

We had to do it via a plug, due to some quirks with `canary`.

Now that `canary` is gone, we can do the more direct route, so we just do it as part of the action itself.

Comment (begedin) [reply lib/code_corps_web/controllers/user_controller.ex]: If you mean the `fetch_session` part, that should not be needed. What happens if you remove it?

Comment (snewcomer) [reply lib/code_corps_web/controllers/user_controller.ex]: @begedin screen shot 2017-09-05 at 7 46 21 am

Comment (begedin) [reply lib/code_corps_web/controllers/user_controller.ex]: @snewcomer Strange, but it makes sense.

I think you actually found a bug in the old login function. We were just calling `Conn.assign` instead of actually formally signing the user in. 

Additionally, it looks like our client is not expecting to sign the user in automatically upon registration and instead does it manually.

Considering these two, I think the simplest thing we can do is outright remove both `fetch_session` and `sign_in`.

Simply put, the code should read `conn |> put_status(:created) |> render("show.json-api", data: user)`and we can leave it that that. 

Sorry to make go through all of this, but hey, we discovered a bug and a piece of code that does nothing along the way!

**Review by begedin — CHANGES_REQUESTED** [review]

File: lib/code_corps/policy/policy.ex [hunk 4]
```diff
@@ -30,6 +30,8 @@ defmodule CodeCorps.Policy do
   defp can?(%User{} = user, :update, %Comment{} = comment, %{}), do: Policy.Comment.update?(user, comment)
   defp can?(%User{} = user, :create, %Organization{}, %{}), do: Policy.Organization.create?(user)
   defp can?(%User{} = user, :update, %Organization{} = organization, %{}), do: Policy.Organization.update?(user, organization)  
+  defp can?(%User{} = current_user, :update, %User{} = user, %{}), do: Policy.User.update?(user, current_user)
+  defp can?(%User{}, _action, %User{} = _user, %{}), do: true
```
Comment (begedin) [inline lib/code_corps/policy/policy.ex]: Remind me again, why is this one needed? It seems unsafe, as it would technically allow any user to perform  any action on any other user. 

If you meant to copy-paste the one from line 46 below, then the third argument would be nil, but that means you'd also have to update the spec to allow nil, not just `User.t` as the third argument.

However, I don't think the catch all from below is strictly necessary when not using canary, so could you double check?

Comment (snewcomer) [reply lib/code_corps/policy/policy.ex]: Yep my bad.  Good call!

**Review by begedin — APPROVED** [review]

**Comment by begedin** [comment]
Hey @snewcomer. I merged! Excellent work.

**Comment by joshsmith** [comment]
🙌  I love seeing things like this happening when I'm not even looking. Great work @snewcomer!
